lib/grandpa/grandpa.go

```diff
@@ -1297,3 +1209,192 @@ func (s *Service) lenVotes(stage Subround) int {
+	authorities []*types.Authority, blockState BlockState) error {
+	if len(fm.Precommits) != len(fm.AuthData) {
+		return ErrPrecommitSignatureMismatch
+	}
+
+	if fm.SetID != setID {
+		return fmt.Errorf("%w: grandpa state set id %d, set id in the commit message %d",
+			ErrSetIDMismatch, setID, fm.SetID)
+	}
+
+	highestFinalizedHeader, err := blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("getting highest finalised header: %w", err)
+	}
+
+	isDescendant, err := blockState.IsDescendantOf(highestFinalizedHeader.Hash(), fm.Vote.Hash)
+	if err != nil {
+		return fmt.Errorf("cannot verify ancestry of highest finalised block: %w", err)
+	}
+
+	if !isDescendant {
+		return errVoteBlockMismatch
+	}
+
+	eqvVoters := getEquivocatoryVoters(fm.AuthData)
+	var count int
+	for i, pc := range fm.Precommits {
+		just := &SignedVote{
+			Vote:        pc,
+			Signature:   fm.AuthData[i].Signature,
+			AuthorityID: fm.AuthData[i].AuthorityID,
+		}
+
+		err := verifyJustification(just, fm.Round, setID, precommit, authorities)
+		if err != nil {
+			logger.Errorf("failed to verify justification for vote from authority id %s, for block hash %s: %s",
+				just.AuthorityID.String(), just.Vote.Hash, err)
+			continue
+		}
+
+		isDescendant, err := blockState.IsDescendantOf(fm.Vote.Hash, just.Vote.Hash)
+		if err != nil {
+			logger.Warnf("could not check for descendant: %s", err)
+			continue
+		}
+
+		precommitedHeader, err := blockState.GetHeader(pc.Hash)
+		if err != nil {
+			return fmt.Errorf("getting header of %s: %w", pc.Hash.Short(), err)
+		}
+
+		if precommitedHeader.Number != uint(pc.Number) {
+			return fmt.Errorf("%w: expected number %d from header but got number %d",
+				ErrBlockHashMismatch, precommitedHeader.Number, pc.Number)
+		}
+
+		if _, ok := eqvVoters[fm.AuthData[i].AuthorityID]; ok {
+			continue
+		}
+
+		if isDescendant {
+			count++
+		}
+	}
+
+	// confirm total # signatures >= grandpa threshold
+	if uint64(count)+uint64(len(eqvVoters)) < threshold {
+		logger.Debugf(
+			"minimum votes not met for finalisation message. Need %d votes and received %d votes.",
+			threshold, count)
+		return ErrMinVotesNotMet
+	}
+
+	logger.Debugf("validated commit message: %v", fm)
+	return nil
+}
+
+func verifyJustification(just *SignedVote, round, setID uint64,
+	stage Subround, authorities []*types.Authority) error {
+	// verify signature
+	msg, err := scale.Marshal(FullVote{
+		Stage: stage,
+		Vote:  just.Vote,
+		Round: round,
+		SetID: setID,
+	})
+	if err != nil {
+		return err
+	}
+
+	pk, err := ed25519.NewPublicKey(just.AuthorityID[:])
+	if err != nil {
+		return err
+	}
+
+	ok, err := pk.Verify(msg, just.Signature[:])
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		return ErrInvalidSignature
+	}
+
+	// verify authority in justification set
+	authFound := false
+
+	for _, auth := range authorities {
+		justKey, err := just.AuthorityID.Encode()
+		if err != nil {
+			return err
+		}
```

Comment:
Move this outside the for loop, no point re-encoding the same thing multiple times.